### PR TITLE
DXDP-842 - Update Styled Components to v5

### PR DESCRIPTION
**Pull request by dmiller9911**
By submitting a PR to this repository, you agree to the terms within the [Auth0 Code of Conduct](https://github.com/auth0/open-source-template/blob/master/CODE-OF-CONDUCT.md). Please see the [contributing guidelines](https://github.com/auth0/.github/blob/master/CONTRIBUTING.md) for how to create and submit a high-quality PR for this repo.

### Description
 
Update to styled-components v5 and fix all breaking changes. This will require a "major" version bump (v0.4.0), but there will be some follow up prs to prevent too many breaking versions.

* Fixed deprecated `withComponent` calls
* `injectGlobal` --> `createGlobalStyle`
* `innerRef` --> `ref`
* Updated react dependency to >=16.8.6 to allow for fix with `forwardRef` and hook calls with shallow rendering

### CODE REVIEW

File: core/components/atoms/_overlay/overlay.tsx [hunk 1]
```diff
@@ -59,68 +60,74 @@ class Overlay extends React.Component<IOverlayProps, IOverlayState> {
...
 
-    if (!this.state.hasBeenMounted) { return null }
+    if (!this.state.hasBeenMounted) {
+      return null;
+    }
 
     const content = open ? (
       <Overlay.Backdrop onMouseDown={this.handleMouseDown}>
-        <Overlay.Element contentSize={contentSize} innerRef={(el) => (this.contentElement = el)}>
+        <Overlay.Element contentSize={contentSize} ref={(el) => (this.contentElement = el)}>
```
Comment (dmiller9911) [inline core/components/atoms/_overlay/overlay.tsx]: The change is here. the rest are prettier changes.

File: core/components/atoms/button/button.tsx [hunk 2]
```diff
@@ -338,20 +340,20 @@ Button.Text = styled.span`
   /* See: https://github.com/auth0/cosmos/pull/947 */
   /* See: https://github.com/auth0/cosmos/pull/1045 */
   margin-top: 1px;
-`
+`;
 
-Button.LinkElement = styled(Button.Element.withComponent('a'))`
+Button.LinkElement = styled(Button.Element).attrs({ as: "a" })`
```
Comment (dmiller9911) [inline core/components/atoms/button/button.tsx]: The change is here. the rest are prettier changes.

File: core/components/atoms/heading/heading.tsx [hunk 3]
```diff
@@ -1,59 +1,60 @@
...
-  )
-}
+  );
+};
 
-Heading.Element = []
+Heading.Element = [];
 
-Heading.Element[1] = styled(BaseHeading.withComponent('h1'))`
+Heading.Element[1] = styled(BaseHeading).attrs({ as: "h1" })`
```
Comment (dmiller9911) [inline core/components/atoms/heading/heading.tsx]: Change here.

File: core/components/atoms/heading/heading.tsx [hunk 4]
```diff
@@ -1,59 +1,60 @@
...
   font-size: 36px;
-`
+`;
 
-Heading.Element[2] = styled(BaseHeading.withComponent('h2'))`
+Heading.Element[2] = styled(BaseHeading).attrs({ as: "h2" })`
```
Comment (dmiller9911) [inline core/components/atoms/heading/heading.tsx]: Change here.

File: core/components/atoms/heading/heading.tsx [hunk 5]
```diff
@@ -1,59 +1,60 @@
...
   font-size: 24px;
   font-weight: ${fonts.weight.medium};
-`
+`;
 
-Heading.Element[3] = styled(BaseHeading.withComponent('h3'))`
+Heading.Element[3] = styled(BaseHeading).attrs({ as: "h3" })`
```
Comment (dmiller9911) [inline core/components/atoms/heading/heading.tsx]: Change here.

File: core/components/atoms/heading/heading.tsx [hunk 6]
```diff
@@ -1,59 +1,60 @@
...
   font-size: 18px; /* TO-DO: tokenize */
   font-weight: ${fonts.weight.bold};
-`
+`;
 
-Heading.Element[4] = styled(BaseHeading.withComponent('h4'))`
+Heading.Element[4] = styled(BaseHeading).attrs({ as: "h4" })`
```
Comment (dmiller9911) [inline core/components/atoms/heading/heading.tsx]: Change here.

File: core/components/atoms/textarea/textarea.tsx [hunk 7]
```diff
@@ -66,7 +65,7 @@ const TextArea = (props: ITextAreaProps) => {
   }
 };
 
-TextArea.Element = styled(StyledInput.withComponent("textarea"))`
+TextArea.Element = styled(StyledInput).attrs({ as: "textarea" })`
```
Comment (dmiller9911) [inline core/components/atoms/textarea/textarea.tsx]: Change here. The rest are prettier changes.

File: core/components/atoms/tooltip/tooltip.tsx [hunk 8]
```diff
@@ -1,58 +1,60 @@
...
   public static ObservedElement = class extends React.Component<IObservedElementProps> {
     public componentDidUpdate() {
-      this.props.scheduleUpdate()
+      this.props.scheduleUpdate();
     }
     public render() {
-      return <Tooltip.Element {...this.props} />
+      const { innerRef, ...props } = this.props;
+
+      return <Tooltip.Element {...props} ref={innerRef} />;
```
Comment (dmiller9911) [inline core/components/atoms/tooltip/tooltip.tsx]: Change here. I kept innerRef for now, but we can switch to forwardRef in a followup.

File: core/components/atoms/tooltip/tooltip.tsx [hunk 9]
```diff
@@ -176,9 +182,9 @@ class Tooltip extends React.Component<ITooltipProps, ITooltipState> {
               onFocus={this.showTooltip}
               onMouseLeave={this.hideTooltip}
               onBlur={this.hideTooltip}
-              innerRef={ref}
+              ref={ref}
```
Comment (dmiller9911) [inline core/components/atoms/tooltip/tooltip.tsx]: Change here.

File: core/components/atoms/tooltip/tooltip.tsx [hunk 10]
```diff
@@ -202,23 +208,19 @@ class Tooltip extends React.Component<ITooltipProps, ITooltipState> {
...
+                  {...Automation("tooltip")}
                   {...props}
                 >
                   {content}
-                  <Tooltip.Arrow
-                    data-placement={placement}
-                    innerRef={arrowProps.ref}
-                    style={arrowProps.style}
-                  />
+                  <Tooltip.Arrow data-placement={placement} ref={arrowProps.ref} style={arrowProps.style} />
```
Comment (dmiller9911) [inline core/components/atoms/tooltip/tooltip.tsx]: Change here.

File: core/components/molecules/dialog/dialog.tsx [hunk 11]
```diff
@@ -110,17 +115,17 @@ const DialogHeader = styled.header`
...
+  const InternalTitle = styled(BaseHeading)`
     font-weight: ${fonts.weight.medium};
     font-size: ${fonts.size.default};
     margin: 0;
-  `
+  `;
 
-  return <InternalTitle {...props} />
-}
+  return <InternalTitle {...props} as={element} />;
```
Comment (dmiller9911) [inline core/components/molecules/dialog/dialog.tsx]: Change here.

File: core/components/molecules/form/field/field.tsx [hunk 12]
```diff
@@ -191,7 +190,8 @@ Field.FieldSetElement = styled.fieldset`
     margin-bottom: ${spacing.medium};
   }
 `;
-Field.CheckboxLabel = StyledLabel.withComponent("legend");
+
+Field.CheckboxLabel = styled(StyledLabel).attrs({ as: "legend" });
```
Comment (dmiller9911) [inline core/components/molecules/form/field/field.tsx]: change here

File: core/components/molecules/list/list.tsx [hunk 13]
```diff
@@ -1,18 +1,14 @@
-import * as React from 'react'
```
Comment (dmiller9911) [inline core/components/molecules/list/list.tsx]: No changes in this file. Only prettier.

File: core/components/styled.ts [hunk 14]
```diff
@@ -29,15 +23,15 @@ domElements.forEach((domElement) => {
...
-  styledWithHelpers[domElement].attrs = styled[domElement].attrs
-})
+  styledWithHelpers[domElement].withConfig = styled[domElement].withConfig;
+  styledWithHelpers[domElement].attrs = styled[domElement].attrs;
+});
 
-export default styledWithHelpers
-export { keyframes, css, injectGlobal, StyledComponentClass, ThemeProvider }
+export default styledWithHelpers;
+export { keyframes, css, createGlobalStyle, ThemeProvider };
```
Comment (dmiller9911) [inline core/components/styled.ts]: change here.

File: internal/docs/component/props.js [hunk 15]
```diff
@@ -15,57 +15,57 @@ const Type = styled.div`
   border-radius: ${misc.radius};
   position: relative;
   left: -${spacing.xsmall};
-`
+`;
 
-const Deprecated = Type.withComponent('span').extend`
+const Deprecated = styled(Type).attrs({ as: "span" })`
```
Comment (dmiller9911) [inline internal/docs/component/props.js]: change here.

File: internal/docs/public/index.html [hunk 16]
```diff
@@ -23,5 +23,5 @@
 
   <body>
     <div id="root"></div>
-  <script type="text/javascript" src="main.9d0573e3d9566df71fc9.js"></script></body>
+  <script type="text/javascript" src="main.71e28e13506319cf9036.js"></script></body>
```
Comment (dmiller9911) [inline internal/docs/public/index.html]: Not sure if these changes are good or bad.

File: internal/test/unit/helpers/list.test.tsx [hunk 17]
```diff
@@ -26,81 +26,76 @@ const testCases = {
...
+    expect(drawer).toBeNull();
+  });
+
+  it("should pass a hidden prop to List.Drawer", () => {
+    const drawer = getDrawer(testCases.includingListDrawer(), false, List.Drawer);
+    expect(drawer[0].props.hidden).toEqual(true);
+  });
+});
+
+describe("isListExpandable", () => {
```
Comment (dmiller9911) [inline internal/test/unit/helpers/list.test.tsx]: Change is in these tests. I was getting some shallow render errors. I reconfigured to make them work.

File: internal/docs/public/index.html (hunk 16 above)
Comment (loginist) [reply internal/docs/public/index.html]: not needed afaik, this should be git ignored

Comment (dmiller9911) [reply internal/docs/public/index.html]: Ok cool. I will add that in a followup.

**Pull request closed without merging**